pulsar-broker/src/test/java/org/apache/pulsar/broker/TopicEventsListenerTest.java

```diff
@@ -0,0 +1,323 @@
+
+    @DataProvider(name = "topicType")
+    public static Object[][] topicType() {
+        return new Object[][] {
+                {"persistent", "partitioned", Boolean.TRUE},
+                {"persistent", "non-partitioned", Boolean.TRUE},
+                {"non-persistent", "partitioned", Boolean.TRUE},
+                {"non-persistent", "non-partitioned", Boolean.TRUE},
+                {"persistent", "partitioned", Boolean.FALSE},
+                {"persistent", "non-partitioned", Boolean.FALSE},
+                {"non-persistent", "partitioned", Boolean.FALSE},
+                {"non-persistent", "non-partitioned", Boolean.FALSE}
+        };
+    }
+
+    @DataProvider(name = "topicTypeNoDelete")
+    public static Object[][] topicTypeNoDelete() {
+        return new Object[][] {
+                {"persistent", "partitioned"},
+                {"persistent", "non-partitioned"},
+                {"non-persistent", "partitioned"},
+                {"non-persistent", "non-partitioned"}
+        };
+    }
+
+    @BeforeClass
+    @Override
+    protected void setup() throws Exception {
+        super.baseSetup();
+        pulsar.getConfiguration().setForceDeleteNamespaceAllowed(true);
+
+        pulsar.getBrokerService().addTopicEventListener((topic, event, stage, t) -> {
+            log.info("got event {}__{} for topic {}", event, stage, topic);
+            if (topic.equals(topicNameToWatch)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("got event {}__{} for topic {} with detailed stack",
+                            event, stage, topic, new Exception("tracing event source"));
+                }
+                events.add(event.toString() + "__" + stage.toString());
+            }
+        });
+    }
+
+    @AfterClass(alwaysRun = true)
+    @Override
+    protected void cleanup() throws Exception {
+        super.internalCleanup();
+    }
+
+    @BeforeMethod
+    protected void setupTest() throws Exception {
+        namespace = "prop/" + UUID.randomUUID();
+        admin.namespaces().createNamespace(namespace, Sets.newHashSet("test"));
+        assertTrue(admin.namespaces().getNamespaces("prop").contains(namespace));
+        admin.namespaces().setRetention(namespace, new RetentionPolicies(3, 10));
+        try (PulsarAdmin admin2 = createPulsarAdmin()) {
+            Awaitility.await().untilAsserted(() ->
+                    assertEquals(admin2.namespaces().getRetention(namespace), new RetentionPolicies(3, 10)));
+        }
+
+        events.clear();
+    }
+
+    @AfterMethod(alwaysRun = true)
+    protected void cleanupTest() throws Exception {
+        deleteNamespaceWithRetry(namespace, true);
+    }
+
+    @Test(dataProvider = "topicType")
+    public void testEvents(Object[] topicType) throws Exception {
+        String topicTypePersistence = (String) topicType[0];
+        String topicTypePartitioned = (String) topicType[1];
+        boolean forceDelete = (Boolean) topicType[2];
```

Comment:
```suggestion
    public void testEvents(String topicTypePersistence, String topicTypePartitioned,
                           boolean forceDelete) throws Exception {
```

And we can use `boolean` instead of `Boolean` for the 3rd array element in `topicType`. The `Object[]` argument can be extracted to variant arguments automatically. You can apply similar changes for other tests as well.